### DISCUSSION TITLE
BM25 + RM3

### DISCUSSION THREAD

**Discussion by lironT74** [castorini/anserini#1976] [Q&A]
Hello,
I have some questions regarding Anserini's implementation of BM25 + RM3 . Disclaimer: I've never used the package, this is asked for comparison purposes. I could not track the answers to my questions in the code.

1. How is p(M_d|q) computed? That is, how does a document's weight in the RM1 model is defined when the initial retrieval is BM25?  
2. After receiving an RM3 model (on top of BM25) what is the approach taken to re-rank the documents? Perhaps CE between the RM3 model and the unsmoothed language models of the documents? 

If you can also point to where this happens in the code so I can try answer any further questions myself that would be great.

Thanks

**Comment by lintool** [comment]
> 1. How is p(M_d|q) computed? That is, how does a document's weight in the RM1 model is defined when the initial retrieval is BM25?

https://github.com/castorini/anserini/blob/master/src/main/java/io/anserini/rerank/lib/Rm3Reranker.java#L83

> 2. After receiving an RM3 model (on top of BM25) what is the approach taken to re-rank the documents? Perhaps CE between the RM3 model and the unsmoothed language models of the documents?

https://github.com/castorini/anserini/blob/master/src/main/java/io/anserini/rerank/lib/Rm3Reranker.java#L85

**Comment by lironT74** [comment]
@lintool Thanks. I would really appreciate further help: 
1. I can see the sum normalization of docs.scores, but what I am missing is whether docs.scores are the raw scores or not? Specifically, if we re-rank initial list received by CE between the query model and the documents language models, the scores are negative, so it doesn't make sense to just sum them, but rather an exp(s) is needed and I cant find where such transformation is done in the code. In my original intend we rerank BM25 and the scores are positive, but still I am interested to know whether there are any transformations (e.g. exp) applied to build p(M_d|q)? 

2. I can see how the feedback query is built, but I am having hard time to find out exactly what approach will be activated [here](https://github.com/castorini/anserini/blob/f0502cc7cf4978b266b6427a594e6c57e524fd8e/src/main/java/io/anserini/rerank/lib/Rm3Reranker.java#L118).

Thank you
